### PR TITLE
Reducing string allocations when reading BasicProperties.

### PR DESCRIPTION
**Pull request by stebet**
## Proposed Changes

Reduces string allocations in BasicProperties by only deserializing the strings when needed.

## Types of Changes

- [X] Bug fix (non-breaking change which fixes issue #965)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause an observable behavior change in existing systems)
- [ ] Documentation improvements (corrections, new content, etc)
- [ ] Cosmetic change (whitespace, formatting, etc)

## Checklist

- [X] I have read the `CONTRIBUTING.md` document
- [X] I have signed the CA (see https://cla.pivotal.io/sign/rabbitmq)
- [X] All tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in related repositories

## Further Comments

This works by storing a reference to the bytes read for header frames in BasicProperties, and lazily reading the strings as they are referenced. For consumer workloads that do not access all the data in BasicProperties, this saves CPU and allocations.

### CODE REVIEW

**Comment by stebet** [comment]
Test app:
```csharp
class Program
{
    private static int messagesSent = 0;
    private static int messagesReceived = 0;
    private static int batchesToSend = 100;
    private static int itemsPerBatch = 5000;
    private static ManualResetEventSlim manualResetEventSlim = new ManualResetEventSlim();
    static async Task Main(string[] args)
    {
        var connectionString = new Uri("amqp://guest:guest@localhost/");
        var connectionFactory = new ConnectionFactory() { DispatchConsumersAsync = false, Uri = connectionString };
        connectionFactory.ClientProvidedName = "BenchMark";
        var connection = connectionFactory.CreateConnection();
        var connection2 = connectionFactory.CreateConnection();
        var publisher = connection.CreateModel();
        var subscriber = connection2.CreateModel();
        publisher.ExchangeDeclare("test", ExchangeType.Topic, true);

        subscriber.QueueDeclare("testqueue", true, false, true);
        var listener = new EventingBasicConsumer(subscriber);
        listener.Received += Listener_Received;
        subscriber.QueueBind("testqueue", "test", "");
        subscriber.BasicConsume("testqueue", true, "testconsumer", listener);
        byte[] payload = new byte[512];

        while (messagesSent < batchesToSend * itemsPerBatch)
        {
            var header = publisher.CreateBasicProperties();
            header.CorrelationId = $"{Guid.NewGuid()}";
            publisher.BasicPublish("test", "", false, header, payload);
            messagesSent++;
        }

        manualResetEventSlim.Wait();

        Console.WriteLine("Done receiving all messages.");
        Console.ReadLine();
        subscriber.Dispose();
        publisher.Dispose();
        connection.Dispose();
        connection2.Dispose();
    }

    private static void Listener_Received(object sender, BasicDeliverEventArgs e)
    {
        Interlocked.Increment(ref messagesReceived);
        if (messagesReceived == batchesToSend * itemsPerBatch)
        {
            manualResetEventSlim.Set();
        }
    }
}
```

**Allocations before:**


**Allocations after:**

**Comment by JanEggers** [comment]
who creates the remaining 1.5 mio strings?

**Comment by stebet** [comment]
> what creates the remaining 1.5 mio strings?

The benchmark app :)

**Comment by JanEggers** [comment]
ahh you are setting the correlationid

**Comment by JanEggers** [comment]
maybe we can pool the BasicProperties on the receiving end and return them when returning the bytearray

File: projects/RabbitMQ.Client/client/framing/BasicProperties.cs [hunk 1]
```diff
@@ -166,69 +284,57 @@ public override string ClusterId
...
+        public override bool IsHeadersPresent() => _headers != default || (_readOnlyMemory.Length > 0 && (_readOnlyMemory.Span[0] & 0b0010_0000) != 0);
 
-        public override bool IsContentEncodingPresent() => _contentEncoding != default;
+        public override bool IsDeliveryModePresent() => _deliveryMode != default || (_readOnlyMemory.Length > 0 && (_readOnlyMemory.Span[0] & 0b0001_0000) != 0);
 
-        public override bool IsHeadersPresent() => _headers != default;
+        public override bool IsPriorityPresent() => _priority != default || (_readOnlyMemory.Length > 0 && (_readOnlyMemory.Span[0] & 0b0000_1000) != 0);
 
-        public override bool IsDeliveryModePresent() => _deliveryMode != default;
+        public override bool IsCorrelationIdPresent() => _correlationId != default || (_readOnlyMemory.Length > 0 && (_readOnlyMemory.Span[0] & 0b0000_0100) != 0);
```
Comment (danielmarbach) [inline projects/RabbitMQ.Client/client/framing/BasicProperties.cs:297]: Can we somehow collapse the hierarchy and thus I line those methods?

File: projects/RabbitMQ.Client/client/framing/BasicProperties.cs [hunk 2]
```diff
@@ -56,85 +61,198 @@ internal sealed class BasicProperties : RabbitMQ.Client.Impl.BasicProperties
...
 
         public override IDictionary<string, object> Headers
         {
-            get => _headers;
+            get
+            {
+                if (_headers == default && IsHeadersPresent())
+                {
+                    WireFormatting.ReadDictionary(_readOnlyMemory.Span.Slice(HeadersOffset), out Dictionary<string, object> tempDirectory);
+                    _headers = tempDirectory;
```
Comment (danielmarbach) [inline projects/RabbitMQ.Client/client/framing/BasicProperties.cs:97]: Temp Directory?

File: projects/RabbitMQ.Client/client/framing/BasicProperties.cs (hunk 1 above)
Comment (stebet) [reply projects/RabbitMQ.Client/client/framing/BasicProperties.cs:297]: I was wondering about that actually, I was thinking that calculating the IsPresent and Offsets to read the properties could be loaded all at once when the first property is accessed to also prevent calculations from being repeated. It should be easy to read and calculate those without allocating on first access.

File: projects/RabbitMQ.Client/client/framing/BasicProperties.cs (hunk 2 above)
Comment (stebet) [reply projects/RabbitMQ.Client/client/framing/BasicProperties.cs:97]: IDictionary<,> vs Dictionary<,> :( Unfortunately we are exposing IDictionary<,> interfaces in a few places and it isn't co/contravariant :(

**Comment by bollhals** [comment]
I know it's not about the test app, but 
`header.CorrelationId = $"{Guid.NewGuid()}";`
Doesn't that allocate 2 strings? (1 for Guid.NewGuid().ToString() and 1 to put the brackets around? => Guid.NewGuid().ToString("B") would get the same output if I'm not mistaken.

I'd be interested in the CPU aspect of this change! The whole "figuring out where my string should start and whether I already have it stored"-part doesn't sound cheap to me.

Downside is that the BasicProperty class got even bigger (+8 MB / 1m objects) and that we pay the prize as well when we have the class to send and not when receiving. 
I like the idea in principal, but I think this would make much more sense if we split the receiving property type from the sending property type.

File: projects/RabbitMQ.Client/client/framing/BasicProperties.cs (hunk 2 above)
Comment (YulerB) [reply projects/RabbitMQ.Client/client/framing/BasicProperties.cs:97]: tempDirectory, can we rename to tempDictionary.

**Comment by stebet** [comment]
> Are you sure? That is, the application will gradually gain memory while reading unique data, such as `correlationid` header.
> 
> https://docs.microsoft.com/en-us/dotnet/api/system.string.intern?view=netcore-3.1#performance-considerations:
> 
> изображение

Spot on. That's why I thought I'd only intern certain strings, which will certainly be duplicated, such as exchange names for example.

File: projects/RabbitMQ.Client/client/framing/BasicProperties.cs (hunk 1 above)
Comment (YulerB) [reply projects/RabbitMQ.Client/client/framing/BasicProperties.cs:297]: It's possible to read all the is present values into a flags enum bypassing all the read bits related code.
For an example:

https://github.com/YulerB/rabbitmq-dotnet-client/blob/master/projects/client/RabbitMQ.Client/src/client/Protocol/BasicProperties.cs

**Comment by stebet** [comment]
> I know it's not about the test app, but
> `header.CorrelationId = $"{Guid.NewGuid()}";`
> Doesn't that allocate 2 strings? (1 for Guid.NewGuid().ToString() and 1 to put the brackets around? => Guid.NewGuid().ToString("B") would get the same output if I'm not mistaken.

It's not adding brackets, it's just string interpolation.

> I'd be interested in the CPU aspect of this change! The whole "figuring out where my string should start and whether I already have it stored"-part doesn't sound cheap to me.

I haven't done a benchmark yet, but I'm willing to bet it's cheaper than doing all the deserialization beforehand. Offsets and presence bits could be precalculated once for all properties on first access.

> Downside is that the BasicProperty class got even bigger (+8 MB / 1m objects) and that we pay the prize as well when we have the class to send and not when receiving.
> I like the idea in principal, but I think this would make much more sense if we split the receiving property type from the sending property type.

Or we utilize object pooling for the BasicProperties, having a pool of them which we can reset after use. That should almost kill the allocations and coupled with lazy loading, could severely reduce the string allocations overhead as well.

Another option would be to create on "Message" object, where the BasicProperties is a property (instead of a separate field, which makes sense to me) and is also lazy loaded. That'd mean that BasicDeliver methods on consumers only have one parameter containing all the information required, and it can be lazy loaded/deserialized from the memory buffers on-demand. This is for example what Kestrel does with incoming HTTP requests. All the information is stored in the HttpRequest object, and lazy-loaded/parsed as it's accessed for maximum efficiency.

**Comment by JanEggers** [comment]
> Another option would be to create on "Message" object
 
+1

**Comment by bollhals** [comment]
> > I know it's not about the test app, but
> > `header.CorrelationId = $"{Guid.NewGuid()}";`
> > Doesn't that allocate 2 strings? (1 for Guid.NewGuid().ToString() and 1 to put the brackets around? => Guid.NewGuid().ToString("B") would get the same output if I'm not mistaken.
> 
> It's not adding brackets, it's just string interpolation.
> 
woops, silly me, sorry.

> Another option would be to create on "Message" object, where the BasicProperties is a property (instead of a separate field, which makes sense to me) and is also lazy loaded. That'd mean that BasicDeliver methods on consumers only have one parameter containing all the information required, and it can be lazy loaded/deserialized from the memory buffers on-demand. This is for example what Kestrel does with incoming HTTP requests. All the information is stored in the HttpRequest object, and lazy-loaded/parsed as it's accessed for maximum efficiency.

I like the idea of this. Especially the reference to the HttpRequest, people are already familiar to this and therefore it will be easily adoptable.

File: projects/RabbitMQ.Client/client/framing/BasicProperties.cs (hunk 2 above)
Comment (michaelklishin) [reply projects/RabbitMQ.Client/client/framing/BasicProperties.cs:97]: @stebet I think folks are simply pointing out that there is a local variable naming typo :)

**Comment by michaelklishin** [comment]
@stebet @bollhals @JanEggers is there anything else that we'd like to do in this PR? Now that we have benchmarking project merged into master, perhaps we can adopt it for this change. Anything else?

**Comment by stebet** [comment]
I'd like to go over it again to be sure about a few things perf-wise, also add a few benchmarks while doing so. I'll update it soon-ish (on paternity leave a.t.m) if that's ok? :)

**Comment by michaelklishin** [comment]
@stebet sure. Sorry to bothering you on your paternity leave 🙏👨 👶 🥇

**Comment by stebet** [comment]
> @stebet sure. Sorry to bothering you on your paternity leave 🙏👨 👶 🥇   

No worries :) I have my notifications turned off, just checking in every now and then.

**Comment by bollhals** [comment]
I have a local change that is based on #962 and extends on #990 , which caches a few common strings like QueueName, ExchangeName, ConsumerTag, AppId. This should make the LRUCache part obsolete.

**Comment by danielmarbach** [comment]
Made a review comment on https://github.com/rabbitmq/rabbitmq-dotnet-client/pull/990/files#r551733728 that maybe influences your work sandro

**Pull request closed without merging**